File: src/features/goals/services/goals.service.ts
import type {
  GoalProgressEntryRow,
  GoalRow,
  GoalStatus,
} from "@/lib/supabase/types"
import { goalsRepository } from "../repositories/goals.repository"
import {
  toGoalSummary,
  validateEntryInput,
  validateGoalInput,
  type GoalEntryInput,
  type GoalFormInput,
  type GoalSummary,
  type GoalValidationCode,
} from "../domain/goals"

export class GoalValidationError extends Error {
  readonly code: GoalValidationCode
  constructor(code: GoalValidationCode) {
    super(code)
    this.code = code
  }
}

export type GoalHistoryEntry = GoalProgressEntryRow & {
  reversedByEntryId: string | null
  replacementEntryId: string | null
}
export type GoalsReadModel = {
  goals: GoalSummary[]
  entriesByGoal: ReadonlyMap<string, GoalHistoryEntry[]>
}
export type GoalHistoryGroup = {
  root: GoalHistoryEntry
  related: GoalHistoryEntry[]
}
export function buildGoalHistoryEntries(
  entries: GoalProgressEntryRow[]
): Map<string, GoalHistoryEntry[]> {
  const reversalByOriginal = new Map(
    entries
      .filter((entry) => entry.reverses_entry_id)
      .map((entry) => [entry.reverses_entry_id!, entry.id])
  )
  const replacementByOriginal = new Map(
    entries
      .filter((entry) => entry.replacement_for_entry_id)
      .map((entry) => [entry.replacement_for_entry_id!, entry.id])
  )
  const entriesByGoal = new Map<string, GoalHistoryEntry[]>()
  for (const entry of entries)
    entriesByGoal.set(entry.goal_id, [
      ...(entriesByGoal.get(entry.goal_id) ?? []),
      {
        ...entry,
        reversedByEntryId: reversalByOriginal.get(entry.id) ?? null,
        replacementEntryId: replacementByOriginal.get(entry.id) ?? null,
      },
    ])
  return entriesByGoal
}
export function groupGoalHistoryEntries(
  entries: GoalHistoryEntry[]
): GoalHistoryGroup[] {
  const byOriginal = new Map<string, GoalHistoryEntry[]>()
  const roots: GoalHistoryEntry[] = []
  for (const entry of entries) {
    const originalId = entry.reverses_entry_id ?? entry.replacement_for_entry_id
    if (originalId) {
      byOriginal.set(originalId, [...(byOriginal.get(originalId) ?? []), entry])
    } else {
      roots.push(entry)
    }
  }
  const collect = (entry: GoalHistoryEntry): GoalHistoryEntry[] =>
    (byOriginal.get(entry.id) ?? [])
      .sort((a, b) => a.created_at.localeCompare(b.created_at))
      .flatMap((child) => [child, ...collect(child)])
  return roots.map((root) => ({ root, related: collect(root) }))
}
export async function loadGoals(): Promise<GoalsReadModel> {
  const data = await goalsRepository.list()
  const entriesByGoal = buildGoalHistoryEntries(data.entries)
  const goals = data.goals.map((goal: GoalRow) => {
    const summary = toGoalSummary(goal, entriesByGoal.get(goal.id) ?? [])
    if (!summary)
      throw new Error(
        "Goal progress is unavailable because stored decimal data is invalid."
      )
    return summary
  })
  return { goals, entriesByGoal }
}
export async function saveGoal(input: GoalFormInput, id?: string) {
  const error = validateGoalInput(input)
  if (error) throw new GoalValidationError(error)
  return id ? goalsRepository.update(id, input) : goalsRepository.create(input)
}
export async function addGoalEntry(goalId: string, input: GoalEntryInput) {
  const error = validateEntryInput(input)
  if (error) throw new GoalValidationError(error)
  return goalsRepository.addEntry(goalId, input)
}
export const correctGoalEntry = goalsRepository.correctEntry
export const setGoalStatus = (id: string, status: GoalStatus) =>
  goalsRepository.setStatus(id, status)
export const setGoalArchived = goalsRepository.setArchived

File: apps/mobile/lib/goals/goals_service.dart
import '../core/data_change.dart';
import 'goal_history.dart';
import 'goal_math.dart';
import 'goal_models.dart';
import 'goals_repository.dart';

/// Thin orchestration over [GoalsRepository] — port of the web
/// `services/goals.service.ts`. Validates form/entry input before it reaches an
/// RPC and pings [DataChange] after every successful mutation so the dashboard
/// Goals card refreshes (the mobile stand-in for the web `tharwati:data-changed`
/// event).
class GoalsService {
  GoalsService([GoalsRepository? repository])
    : _repo = repository ?? GoalsRepository();

  final GoalsRepository _repo;

  /// Every goal summarised, plus each goal's history entries with correction
  /// back-links (port of `loadGoals`).
  Future<GoalsReadModel> loadGoals() async {
    final data = await _repo.list();
    final entriesByGoal = buildGoalHistoryEntries(data.entries);

    final goals = data.goals.map((goal) {
      final summary = toGoalSummary(
        goal,
        (entriesByGoal[goal.id] ?? const []).map((h) => h.entry).toList(),
      );
      if (summary == null) {
        throw StateError(
          'Goal progress is unavailable — stored decimal data is invalid.',
        );
      }
      return summary;
    }).toList();

    return GoalsReadModel(goals: goals, entriesByGoal: entriesByGoal);
  }

  Future<void> saveGoal(GoalFormInput input, {String? id}) async {
    final error = validateGoalInput(input);
    if (error != null) throw GoalActionException(goalValidationMessage(error));
    if (id == null) {
      await _repo.createGoal(input);
    } else {
      await _repo.updateGoal(id, input);
    }
    DataChange.instance.ping();
  }

  Future<void> addGoalEntry(String goalId, GoalEntryInput input) async {
    final error = validateEntryInput(input);
    if (error != null) throw GoalActionException(goalValidationMessage(error));
    await _repo.addEntry(goalId, input);
    DataChange.instance.ping();
  }

  Future<void> correctGoalEntry(
    String entryId, {
    String? amount,
    String? effectiveOn,
    String? note,
  }) async {
    await _repo.correctEntry(
      entryId,
      amount: amount,
      effectiveOn: effectiveOn,
      note: note,
    );
    DataChange.instance.ping();
  }

  Future<void> setGoalStatus(String goalId, String status) async {
    await _repo.setStatus(goalId, status);
    DataChange.instance.ping();
  }

  Future<void> setGoalArchived(String goalId, bool archived) async {
    await _repo.setArchived(goalId, archived);
    DataChange.instance.ping();
  }
}
